### PR TITLE
Autocompletion

### PR DESCRIPTION
**Pull request by ulrikeschaefer**
## Basics

These points need to be fulfilled for every PR:

- [x ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. - nothing needs got get built
- [x ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

File: src/tools/autocompletion/find_autocompletion_options.sh [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#/usr/bin/env bash
+
+_helloworld_completions() {
+	# completing for last input before a tab
+	OUTPUT="$(python3 find_autocompletion_options.py ${COMP_WORDS[-1]})"
```
Comment (markus2330) [inline src/tools/autocompletion/find_autocompletion_options.sh]:
```suggestion
	OUTPUT="$(python3 find_autocompletion_options.py spec/autocomplete ${COMP_WORDS[-1]})"
```

here you could pass the path to the specification

File: src/tools/autocompletion/find_autocompletion_options.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import kdb
+import sys
+
+start_of_word = None
+
+def get_start_of_word():
+	global start_of_word
+	if (len(sys.argv) > 1):
+		start_of_word = sys.argv[1]
```
Comment (markus2330) [inline src/tools/autocompletion/find_autocompletion_options.py]:
```suggestion
		start_of_word = sys.argv[2]
```

File: src/tools/autocompletion/find_autocompletion_options.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	global start_of_word
+	start_of_word = input_start_of_word
+	return find_auto_completion_options()
+
+
+def find_auto_completion_options():
+	global start_of_word
+	with kdb.KDB() as k:
+		ks = kdb.KeySet()
+		k.get(ks, 'spec/autocomplete')
```
Comment (markus2330) [inline src/tools/autocompletion/find_autocompletion_options.py]:
```suggestion
		k.get(ks, sys.argv[1])
```

File: src/tools/autocompletion/find_autocompletion_options.sh (hunk 1 above)
Comment (ulrikeschaefer) [reply src/tools/autocompletion/find_autocompletion_options.sh]: I have added `-m spec/tests/autocomplete`

File: src/tools/autocompletion/README.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# Autocompletion
+
+To get autocompletion to run you need to have python3 installed and the kdb-python Module up and running.
```
Comment (markus2330) [inline src/tools/autocompletion/README.md]: link to python binding

File: src/tools/autocompletion/README.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+## Bash
+
+## How to get it to run
+To get the bash completion to run you need to run `source find_autocompletion_options.sh`. After that in your console you should be able to type `kdb <TAB><TAB>` and have completion options show up.
```
Comment (markus2330) [inline src/tools/autocompletion/README.md]: Mounting of specification not present. Error handling for root key not present. I currently get:

```
markus@zbook:~/Projekte/Elektra/current/src/tools/autocompletion$ kdb Traceback (most recent call last):
  File "find_autocompletion_options.py", line 190, in <module>
    result = find_auto_completion_options()
  File "find_autocompletion_options.py", line 60, in find_auto_completion_options
    root = root_key.getMeta('root').value
AttributeError: 'NoneType' object has no attribute 'getMeta'
```

File: src/tools/autocompletion/README.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+## How to test
```
Comment (markus2330) [inline src/tools/autocompletion/README.md]: How to Run Tests

File: src/tools/autocompletion/README.md [hunk 7]
```diff
@@ -0,0 +1,12 @@
+# Autocompletion
+
+To get autocompletion to run you need to have python3 installed and the kdb-python Module up and running.
+
+## Bash
+
+## How to get it to run
+To get the bash completion to run you need to run `source find_autocompletion_options.sh`. After that in your console you should be able to type `kdb <TAB><TAB>` and have completion options show up.
+
```
Comment (markus2330) [inline src/tools/autocompletion/README.md]: If I do `kdb mount $(pwd)/spec spec/tests/autocomplete ni` I get an error when typing `kdb list-tools <tab>`:

```
markus@zbook:~/Projekte/Elektra/current/src/tools/autocompletion$ kdb list-tools Traceback (most recent call last):
  File "find_autocompletion_options.py", line 190, in <module>
    result = find_auto_completion_options()
  File "find_autocompletion_options.py", line 66, in find_auto_completion_options
    option_completion = complete_options(key_last_word)
  File "find_autocompletion_options.py", line 141, in complete_options
    len_opts = int((key.getMeta(name='opt').value)[1:])
AttributeError: 'NoneType' object has no attribute 'value'
```

(kdb ls tab works)

**Comment by markus2330** [comment]
Looks good, we can slowly merge it.

File: src/tools/autocompletion/erm-spec [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+description = explain what is being done
+
+[showversion]
+opt/long = version
+opt/arg = none
+description = output version information and exit
+
+[files/#]
+args = remaining
+env = FILES
```
Comment (ulrikeschaefer) [inline src/tools/autocompletion/erm-spec]: mounted with:
sudo kdb mount ~/libelektra/src/tools/autocompletion/erm-spec spec/tests/autocomplete/erm -f ni

File: src/tools/autocompletion/find_autocompletion_options.py [hunk 9]
```diff
@@ -137,12 +139,56 @@ def complete_commands(start_of_input):
...
+		if meta_opt and meta_opt.value.startswith('#'):
+			len_opts = int((meta_opt.value)[1:])
+			i = 0
+			while i <= len_opts:
+				i+=1
+				opt = key.getMeta(name='opt/#{}'.format(i))
+				opt_long = key.getMeta(name='opt/#{}/long'.format(i))
+				if start_of_current_input is None or len_of_word == 0:
+					if opt_long:
+
```
Comment (ulrikeschaefer) [inline src/tools/autocompletion/find_autocompletion_options.py]: currently as a result I get -I and -R
so only opt for the last element in the list, when printing opt_long, it is always None. Could this be a problem with how my spec is mounted?

**Comment by markus2330** [comment]
Can you also post the stacktrace here?

File: src/tools/autocompletion/find_autocompletion_options.py [hunk 10]
```diff
@@ -137,12 +139,56 @@ def complete_commands(start_of_input):
...
+	if start_of_current_input is not None:
+		start_of_word = start_of_current_input.replace('-', '')
+		len_of_word = len(start_of_current_input.strip())
+	k = kdb.KDB()
+	ks = kdb.KeySet()
+	k.get(ks, mount_point)
+	for key in ks:
+		path = str(key).split('/')
+		meta_opt = key.getMeta(name='opt')
+		if meta_opt and meta_opt.value.startswith('#'):
```
Comment (markus2330) [inline src/tools/autocompletion/find_autocompletion_options.py]: for non-arrays it does not start with `#`.

File: src/tools/autocompletion/find_autocompletion_options.py [hunk 11]
```diff
@@ -137,12 +139,56 @@ def complete_commands(start_of_input):
...
-	global start_of_current_input, last_command
+	global last_command
+	completion_options = []
+	if last_command is not None:
+		completion_options.extend(complete_options_opt_array(last_command))
+	else:
+		completion_options.extend(complete_options_single())
+	return completion_options
+
+def complete_options_single():
```
Comment (markus2330) [inline src/tools/autocompletion/find_autocompletion_options.py]: Please document what this function should do: should it handle both single options in array and non-arrays?

File: src/tools/autocompletion/find_autocompletion_options.py (hunk 9 above)
Comment (markus2330) [reply src/tools/autocompletion/find_autocompletion_options.py]: You can always check if everything is as expected by using `kdb meta-ls` and `kdb meta-get`, e.g. `kdb meta-ls spec/tests/autocomplete/erm/force` or `kdb meta-get spec/tests/autocomplete/erm/force opt`

File: src/tools/autocompletion/README.md [hunk 12]
```diff
@@ -0,0 +1,15 @@
+# Autocompletion
+
+To get autocompletion to run you need to have python3 installed and the kdb-python Module up and running.
+For the kdb-python Module to run, the `~/.bashrc` needs to contain `export PYTHONPATH=${PYTHONPATH}:/PATH/TO/PYTHON`.
+
+## Bash
+
+## How to get it to run
+First you need to mount the specification file with kdb by running `sudo kdb mount ~/PATH/TO/SPEC/FILE/spec_ spec/tests/autocomplete -f ni`.
+To get the bash completion to run you need to run `source find_autocompletion_options.sh`. The specifiaction File that should be used needs to be mounted
```
Comment (sanssecours) [inline src/tools/autocompletion/README.md]:
```suggestion
To get the bash completion to run you need to run `source kdb_find_autocompletion_otpions.sh`. The specification file that should be used needs to be mounted
```

File: src/tools/autocompletion/README.md [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+After that in your console you should be able to type `kdb <TAB><TAB>` and have completion options show up.
+
+## How to Run Tests
+Currently it is necessary to call the test-script from the folder the `find_autocompletion_options.py` is located.
```
Comment (sanssecours) [inline src/tools/autocompletion/README.md]:
```suggestion
Currently it is necessary to call the test-script from the folder where `find_autocompletion_options.py` is located.
```

File: src/tools/autocompletion/find_autocompletion_options.py [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+	global mount_point, root, last_command, typed, start_of_current_input
+	mount_point = input_mount_point
+	root = input_root
+	start_of_current_input = input_start_of_current_input
+	last_command = input_last_command
+	typed = input_typed
+	return find_auto_completion_options()
+
+# depending on last_word calls completion for commands or arguments
+# return: string seperated by a \n, every line is one possible completion
```
Comment (sanssecours) [inline src/tools/autocompletion/find_autocompletion_options.py]:
```suggestion
# return: string separated by a \n, every line is one possible completion 
```

File: src/tools/autocompletion/find_autocompletion_options.py [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+			else:
+				if opt_long:
+					completion_options.append('--' + opt_long.value)
+				elif opt:
+					completion_options.append('-' + opt.value)
+		k.close()
+	return completion_options
+
+# input: command, a string that should be a shell command
+#		seperator, a char that seperates every result in the result
```
Comment (sanssecours) [inline src/tools/autocompletion/find_autocompletion_options.py]:
```suggestion
#		separator, a char that separates every result in the result 
```

File: src/tools/autocompletion/erm_find_autocompletion_otpions.sh [hunk 16]
```diff
@@ -0,0 +1,18 @@
+#/usr/bin/env bash
+IFS=$'\n'
+
+_erm_find_completions() {
+	COMPREPLY=()
+	local cur="${COMP_WORDS[COMP_CWORD]}"
+	local cur_str=""
+	if ! [ -z "${cur}" ]; then
```
Comment (sanssecours) [inline src/tools/autocompletion/erm_find_autocompletion_otpions.sh]:
```suggestion
	if [ -n "${cur}" ]; then
```

File: src/tools/autocompletion/kdb_find_autocompletion_otpions.sh [hunk 17]
```diff
@@ -0,0 +1,18 @@
+#/usr/bin/env bash
+IFS=$'\n'
+
+_kdb_find_completions() {
+	COMPREPLY=()
+	local cur="${COMP_WORDS[COMP_CWORD]}"
+	local cur_str=""
+	if ! [ -z "${cur}" ]; then
```
Comment (sanssecours) [inline src/tools/autocompletion/kdb_find_autocompletion_otpions.sh]:
```suggestion
	if [ -n "${cur}" ]; then
```

File: src/tools/autocompletion/README.md [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+From this folder call `python3 test_SOME_TEST.py`.
```
Comment (sanssecours) [inline src/tools/autocompletion/README.md]:
```suggestion
From this folder call `python3 test_find_autocompletion_commands_and_opts.py`.
```

On my machine some of the tests fail:

```
TESTCASE: all
RESULTS:
['--all', '--color', '--color', '--color', '--color', '--color', '--debug', '--debug', '--debug', '--debug', '--load', '--max-depth', '--min-depth', '--namespace', '--no-newline', '--null', '--null', '--plugins-config', '--profile', '--profile', '--profile', '--profile', '--profile', '--quiet', '--verbose', '--verbose', '--verbose', '--version', '--version', '--version', '--version', '--version', '-0', '-0', '-C', '-C', '-C', '-C', '-C', '-M', '-N', '-V', '-V', '-V', '-V', '-V', '-a', '-c', '-d', '-d', '-d', '-d', '-l', '-m', '-n', '-p', '-p', '-p', '-p', '-p', '-q', '-v', '-v', '-v']
EXPECTED RESULTS:
['get', 'list-tools', 'ls', 'plugin-info', 'plugin-list', 'set']
DIFFERENT LENGTHS OF RESULT & EXPECTED RESULT
LENGHT OF RESULT: 64
LENGHT OF EXPECTED RESULT: 6
MISSING: get
MISSING: list-tools
MISSING: ls
MISSING: plugin-info
MISSING: plugin-list
MISSING: set
NOT EXPECTED: --all
NOT EXPECTED: --color
NOT EXPECTED: --color
NOT EXPECTED: --color
NOT EXPECTED: --color
NOT EXPECTED: --color
NOT EXPECTED: --debug
NOT EXPECTED: --debug
NOT EXPECTED: --debug
NOT EXPECTED: --debug
NOT EXPECTED: --load
NOT EXPECTED: --max-depth
NOT EXPECTED: --min-depth
NOT EXPECTED: --namespace
NOT EXPECTED: --no-newline
NOT EXPECTED: --null
NOT EXPECTED: --null
NOT EXPECTED: --plugins-config
NOT EXPECTED: --profile
NOT EXPECTED: --profile
NOT EXPECTED: --profile
NOT EXPECTED: --profile
NOT EXPECTED: --profile
NOT EXPECTED: --quiet
NOT EXPECTED: --verbose
NOT EXPECTED: --verbose
NOT EXPECTED: --verbose
NOT EXPECTED: --version
NOT EXPECTED: --version
NOT EXPECTED: --version
NOT EXPECTED: --version
NOT EXPECTED: --version
NOT EXPECTED: -0
NOT EXPECTED: -0
NOT EXPECTED: -C
NOT EXPECTED: -C
NOT EXPECTED: -C
NOT EXPECTED: -C
NOT EXPECTED: -C
NOT EXPECTED: -M
NOT EXPECTED: -N
NOT EXPECTED: -V
NOT EXPECTED: -V
NOT EXPECTED: -V
NOT EXPECTED: -V
NOT EXPECTED: -V
NOT EXPECTED: -a
NOT EXPECTED: -c
NOT EXPECTED: -d
NOT EXPECTED: -d
NOT EXPECTED: -d
NOT EXPECTED: -d
NOT EXPECTED: -l
NOT EXPECTED: -m
NOT EXPECTED: -n
NOT EXPECTED: -p
NOT EXPECTED: -p
NOT EXPECTED: -p
NOT EXPECTED: -p
NOT EXPECTED: -p
NOT EXPECTED: -q
NOT EXPECTED: -v
NOT EXPECTED: -v
NOT EXPECTED: -v
FAILURE - all
TESTCASE: start with: a
RESULTS:
['--all', '-a']
EXPECTED RESULTS:
[]
DIFFERENT LENGTHS OF RESULT & EXPECTED RESULT
LENGHT OF RESULT: 2
LENGHT OF EXPECTED RESULT: 0
NOT EXPECTED: --all
NOT EXPECTED: -a
FAILURE - start with: a
TESTCASE: start with: b
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: b
TESTCASE: start with: c
RESULTS:
['--color', '--color', '--color', '--color', '--color', '-c']
EXPECTED RESULTS:
[]
DIFFERENT LENGTHS OF RESULT & EXPECTED RESULT
LENGHT OF RESULT: 6
LENGHT OF EXPECTED RESULT: 0
NOT EXPECTED: --color
NOT EXPECTED: --color
NOT EXPECTED: --color
NOT EXPECTED: --color
NOT EXPECTED: --color
NOT EXPECTED: -c
FAILURE - start with: c
TESTCASE: start with: d
RESULTS:
['--debug', '--debug', '--debug', '--debug', '-d', '-d', '-d', '-d']
EXPECTED RESULTS:
[]
DIFFERENT LENGTHS OF RESULT & EXPECTED RESULT
LENGHT OF RESULT: 8
LENGHT OF EXPECTED RESULT: 0
NOT EXPECTED: --debug
NOT EXPECTED: --debug
NOT EXPECTED: --debug
NOT EXPECTED: --debug
NOT EXPECTED: -d
NOT EXPECTED: -d
NOT EXPECTED: -d
NOT EXPECTED: -d
FAILURE - start with: d
TESTCASE: start with: e
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: e
TESTCASE: start with: f
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: f
TESTCASE: start with: g
RESULTS:
[]
EXPECTED RESULTS:
['get']
DIFFERENT LENGTHS OF RESULT & EXPECTED RESULT
LENGHT OF RESULT: 0
LENGHT OF EXPECTED RESULT: 1
MISSING: get
FAILURE - start with: g
TESTCASE: start with: h
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: h
TESTCASE: start with: i
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: i
TESTCASE: start with: j
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: j
TESTCASE: start with: k
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: k
TESTCASE: start with: l
RESULTS:
['--load', '-l']
EXPECTED RESULTS:
['list-tools', 'ls']
MISSING: list-tools
MISSING: ls
NOT EXPECTED: --load
NOT EXPECTED: -l
FAILURE - start with: l
TESTCASE: start with: m
RESULTS:
['--max-depth', '--min-depth', '-m']
EXPECTED RESULTS:
[]
DIFFERENT LENGTHS OF RESULT & EXPECTED RESULT
LENGHT OF RESULT: 3
LENGHT OF EXPECTED RESULT: 0
NOT EXPECTED: --max-depth
NOT EXPECTED: --min-depth
NOT EXPECTED: -m
FAILURE - start with: m
TESTCASE: start with: n
RESULTS:
['--namespace', '--no-newline', '--null', '--null', '-n']
EXPECTED RESULTS:
[]
DIFFERENT LENGTHS OF RESULT & EXPECTED RESULT
LENGHT OF RESULT: 5
LENGHT OF EXPECTED RESULT: 0
NOT EXPECTED: --namespace
NOT EXPECTED: --no-newline
NOT EXPECTED: --null
NOT EXPECTED: --null
NOT EXPECTED: -n
FAILURE - start with: n
TESTCASE: start with: o
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: o
TESTCASE: start with: p
RESULTS:
['--plugins-config', '--profile', '--profile', '--profile', '--profile', '--profile', '-p', '-p', '-p', '-p', '-p']
EXPECTED RESULTS:
['plugin-info', 'plugin-list']
DIFFERENT LENGTHS OF RESULT & EXPECTED RESULT
LENGHT OF RESULT: 11
LENGHT OF EXPECTED RESULT: 2
MISSING: plugin-info
MISSING: plugin-list
NOT EXPECTED: --plugins-config
NOT EXPECTED: --profile
NOT EXPECTED: --profile
NOT EXPECTED: --profile
NOT EXPECTED: --profile
NOT EXPECTED: --profile
NOT EXPECTED: -p
NOT EXPECTED: -p
NOT EXPECTED: -p
NOT EXPECTED: -p
NOT EXPECTED: -p
FAILURE - start with: p
TESTCASE: start with: q
RESULTS:
['--quiet', '-q']
EXPECTED RESULTS:
[]
DIFFERENT LENGTHS OF RESULT & EXPECTED RESULT
LENGHT OF RESULT: 2
LENGHT OF EXPECTED RESULT: 0
NOT EXPECTED: --quiet
NOT EXPECTED: -q
FAILURE - start with: q
TESTCASE: start with: r
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: r
TESTCASE: start with: s
RESULTS:
[]
EXPECTED RESULTS:
['set']
DIFFERENT LENGTHS OF RESULT & EXPECTED RESULT
LENGHT OF RESULT: 0
LENGHT OF EXPECTED RESULT: 1
MISSING: set
FAILURE - start with: s
TESTCASE: start with: t
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: t
TESTCASE: start with: u
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: u
TESTCASE: start with: v
RESULTS:
['--verbose', '--verbose', '--verbose', '--version', '--version', '--version', '--version', '--version', '-v', '-v', '-v']
EXPECTED RESULTS:
[]
DIFFERENT LENGTHS OF RESULT & EXPECTED RESULT
LENGHT OF RESULT: 11
LENGHT OF EXPECTED RESULT: 0
NOT EXPECTED: --verbose
NOT EXPECTED: --verbose
NOT EXPECTED: --verbose
NOT EXPECTED: --version
NOT EXPECTED: --version
NOT EXPECTED: --version
NOT EXPECTED: --version
NOT EXPECTED: --version
NOT EXPECTED: -v
NOT EXPECTED: -v
NOT EXPECTED: -v
FAILURE - start with: v
TESTCASE: start with: w
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: w
TESTCASE: start with: x
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: x
TESTCASE: start with: y
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: y
TESTCASE: start with: z
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: z
TESTCASE: start with: lis
RESULTS:
[]
EXPECTED RESULTS:
['list-tools']
DIFFERENT LENGTHS OF RESULT & EXPECTED RESULT
LENGHT OF RESULT: 0
LENGHT OF EXPECTED RESULT: 1
MISSING: list-tools
FAILURE - start with: lis
TESTCASE: start with: rm
RESULTS:
[]
EXPECTED RESULTS:
[]
SUCCESS - start with: rm
TESTCASE: start with: plugi
RESULTS:
['--plugins-config']
EXPECTED RESULTS:
['plugin-info', 'plugin-list']
DIFFERENT LENGTHS OF RESULT & EXPECTED RESULT
LENGHT OF RESULT: 1
LENGHT OF EXPECTED RESULT: 2
MISSING: plugin-info
MISSING: plugin-list
NOT EXPECTED: --plugins-config
FAILURE - start with: plugi

Tests run: 30
FAILURES: 14
SUCCESSES: 16
```

. Is this expected behavior?

File: src/tools/autocompletion/README.md [hunk 19]
```diff
@@ -0,0 +1,15 @@
+# Autocompletion
+
+To get autocompletion to run you need to have python3 installed and the kdb-python Module up and running.
+For the kdb-python Module to run, the `~/.bashrc` needs to contain `export PYTHONPATH=${PYTHONPATH}:/PATH/TO/PYTHON`.
+
+## Bash
+
+## How to get it to run
+First you need to mount the specification file with kdb by running `sudo kdb mount ~/PATH/TO/SPEC/FILE/spec_ spec/tests/autocomplete -f ni`.
```
Comment (sanssecours) [inline src/tools/autocompletion/README.md]:
```suggestion
First you need to mount the specification file with kdb by running `sudo kdb mount /ABSOLUTE/PATH/TO/SPEC/FILE spec/tests/autocomplete -f ni`.
```

Comment (markus2330) [reply src/tools/autocompletion/README.md]: I would prefer if you say that you assume that someone is in the folder and then use `$(pwd)/spec` to refer to the spec file.

Comment (markus2330) [reply src/tools/autocompletion/README.md]: (Of course you need to explicitly write down that we assume someone is in this folder.)

**Comment by markus2330** [comment]
Can you maybe close these two PRs (also the restyled) and create a new one (which is already restyled)?

**Comment by markus2330** [comment]
@kodebach can you maybe rebase this and have a look what is missing for the complete command-line spec we currently have?

@ulrikeschaefer wants to continue the work on this and it is probably easier for her if we get this merged before.

**Comment by kodebach** [comment]
1. I don't really want to rebase the `master` branch of another user's fork (if I even have the rights to do that). @ulrikeschaefer should probably close the PR, create a new branch in their fork, copy/cherry-pick the changes and open a new PR.
2. The only conflict is in the release notes. That shouldn't be too difficult to rebase. If all else fails, take the master version during the rebase and re-add the new changelog entry afterwards.
3. I haven't looked too closely at the code yet, but AFAICT the main thing that needs to be changed is that this PR uses a completely different syntax for (sub-)commands compared to `elektraGetOpts`.

**Pull request closed without merging**